### ISSUE TITLE
overlapping hooks aren't very robust

### ISSUE THREAD

**Issue by lunixbochs** [unicorn-engine/unicorn#375]
Different projects using Unicorn (or different components of the same project) can't overlap very well right now because many of the hook forms expect to be the only hook installed.

The only exception I know is hooks with range parameters, of which you can have multiple, as long as you don't install a hook with no range parameters.

This isn't very powerful, and doesn't lend itself to making components that extend Unicorn nicely. If I made a code library that implements stack tracing internally, it's much nicer for it to maintain its own Unicorn hook than to centralize everything in my code around a single code hook, a single memory hook, etc.

I'm actually already running into this in Usercorn, where I have a main instance that manages a binary, runs it, handles syscalls, etc, but I also have a networked debugger that wants to attach additional hooks. It'd be nice if I could just pile more hooks on top of Unicorn, but that's not really an option right now.

----

TLDR: I think the hook system needs some love. The success of `hook_add()` should never depend on existing hooks.

**Comment by williballenthin** [comment]
+1

my project have its own hook multiplexer that works well within my project, but would not work well with other libraries that ignore its existence.

**Comment by farmdve** [comment]
Sure I agree that the hook subsystem needs some revamping, but I am sure you can all agree that Unicorn needs more people contributing.

I know that you have Usercorn to work on, but I have my own little project too. This is why I am working on Unicorn. I wanted post-instruction callback, I implemented it, I wanted less leaks, so I fixed them. I want the memory read hook to return a value, so I fixed that too. I wanted to change EIP during callback, I added that too(see https://github.com/farmdve/unicorn/commit/bd2f11117b116ba25a3f439a60b257015437fe0a) and then I needed speed, I also improved that.

**Comment by lunixbochs** [comment]
If you're trying to insinuate/suggest something [you might](https://github.com/unicorn-engine/unicorn/commits?author=lunixbochs) have the [wrong idea](https://github.com/unicorn-engine/unicorn/pulls?q=is%3Apr+author%3Alunixbochs+is%3Aclosed)

I was planning on working on this issue later, but I had the idea during lunch at work and didn't have immediate time so I logged it.

**Comment by aquynh** [comment]
yes this should be fixed. pull requests are welcome.

while we are at it, i think we can make all kind of hooks to support address range. for example, you might want to hook `UC_HOOK_INTR` or `UC_HOOK_INSN`, but only in a memory range [0x1000, 0x2000].
same goes for **invalid** memory access events `UC_HOOK_MEM_READ_UNMAPPED`, `UC_HOOK_MEM_READ_PROT`, etc.

**Comment by lunixbochs** [comment]
Okay, that part should be straightforward. I'll consider it during implementation.

These are my thoughts / what I've been working on:

Each hook type will have a linked list, where elements contain a hook ID and a pointer to a hook data struct.

To add a hook, we first add an element to a global hook list, then use its address in memory as the hook ID stored in each specific list. This allows multiple hook types added at once to be later deleted properly, and makes for easy bookkeeping.

To delete a hook, we walk the global list to confirm it exists and get its ID, then remove that ID from each hook type's linked list.

When calling a hook type, we can check if `list->head != NULL`, after which all hooks in the list will be called (if they match any address filters). In the special case of invalid memory hooks, if one hook returns True the remaining hooks will not be called.

The instruction counting hook will be automatically added and removed as necessary in `uc_emu_start()`. The `UC_ERR_HOOK_EXIST` enum will be removed.

**Comment by lunixbochs** [comment]
Also, if we're going to add begin/end to everything I guess it shouldn't be varargs anymore.

**Comment by aquynh** [comment]
yes, all sound good to me!

**Comment by lunixbochs** [comment]
Alright, this is a kinda huge patch.

WIP code here (not fully working yet): https://github.com/lunixbochs/unicorn/commit/0df47d0d173e27fa67a0504e5388390be34d2503

Thoughts / basic code review welcome. I think I caught a couple of minor bugs in the existing code as well.

**Comment by aquynh** [comment]
a quick comment: how about renaming `llist.h` to `list.h` or even `link_list.h`, or something more readable? 

same goes for struct/vars llist_xxx

**Comment by mike820324** [comment]
I'm curious if using link list, how to properly handle the return value of  hook functions. For example, 

- ```uc_cb_eventmem_t```  will depend on the return value to decide whether to continue emulation.

> Even though, currently, there is only one hook type that will have return value, others are just a stateless callback.

If using a link list, the user may need to setup the hook callbacks in the correct order, or the behavior may not be what they want.

**Comment by aquynh** [comment]
@mike820324 you are right, but i dont think there is a perfect solution for the problem of callback priority. i guess all callback systems have the same issue.

**Comment by aquynh** [comment]
@lunixbochs, now look closer, i have some more comments:

- you said "In the special case of invalid memory hooks, if one hook returns True the remaining hooks will not be called". how about executing all the callbacks either until the last one, or until the first one return False?

- what do you mean by "The instruction counting hook will be automatically added and removed..."? this is what we are already doing now, no?

thanks.

**Comment by lunixbochs** [comment]
> this is what we are already doing now, no?

No, you were transparently enabling it in the hook_find() function. We don't have a hook_find() function anymore, so I'm actually adding it to the linked list.

> how about executing all the callbacks either until the last one, or until the first one return False?

No, because it's unlikely every single callback will map the memory for us. `true` means the memory was mapped/approved by the callback. `false` means the CPU should exit.

The use case here is something like: I create two callbacks that can map memory automatically in some situations. If the first callback fails, (returns false), the second callback might still be able to map the memory. This means we should stop the callback chain if any callback maps the memory and returns `true`, because we no longer have invalid memory.

**Comment by aquynh** [comment]
it is tricky to handle callbacks for invalid memory events. apparently we only support maximum 1 "false" and 1 "true" callback. i am wondering if there is any way to make this a policy & enforce it to keep everything more simple

> How about ulist / ulist.h, or uclist?

now i am trying to guess what "ulist" and "uclist" mean, so this does not seem good enough.

how about simply "list", because linked list is the only type of list we need?

**Comment by lunixbochs** [comment]
>  apparently we only support maximum 1 "false" and 1 "true" callback

That doesn't need to be the case. I think `false` should mean "continue with normal operation", which usually means a crash, but could mean calling the next callback. `true` means "we handled this exceptional condition, so you don't need to call any more callbacks"

**Comment by lunixbochs** [comment]
Seems to be working now except for a sporadic hang that's now developed.

**Comment by lunixbochs** [comment]
Alright, hang fixed (was a mistake when renaming the list struct causing out of bounds access). Seems to be working with all of my tests now, so I'm going to PR.

**Comment by sparktrend** [comment]
I can't wait to see this feature especially with the memory range selection!

**Comment by aquynh** [comment]
@sparktrend: see https://github.com/unicorn-engine/unicorn/issues/430. if you find any problems, please report. thanks.
